borg/archiver.py

```diff
@@ -353,6 +355,155 @@ def do_extract(self, args):
+                # This is a link
+                if item.get(b"source", None) != compare_item.get(b"source", None):
+                    if args.json:
+                        dump_json({
+                            "path": path,
+                            "what": "link",
+                            "ref": item.get(b"source", "<regular file>"),
+                            "cmp": compare_item.get(b"source", "<regular file>")
+                        })
+                    else:
+                        print(remove_surrogates(path), "different link")
+                        print("\t", args.location.archive, item.get(b'source', '<regular file>'))
+                        print("\t", args.compare, compare_item.get(b'source', '<regular file>'))
+                return
+            if item[b'chunks'] != compare_item[b'chunks'] or deleted or not can_compare_chunks:
+                # Contents are different
+                ref_chunks = {c[0] for c in item[b'chunks']}
+                cmp_chunks = {c[0] for c in compare_item[b'chunks']}
+                total_ref = None if not item[b'mode'] else sum(c[1] for c in item[b'chunks'])
+                total_compare = None if not compare_item[b'mode'] else sum(c[1] for c in compare_item[b'chunks'])
+                if (not can_compare_chunks and total_ref == total_compare and not deleted and
+                        compare_chunks(ref_chunks, cmp_chunks, ref, compare)):
+                    return
+                added = sum(c[1] for c in (cmp_chunks - ref_chunks))
+                removed = sum(c[1] for c in (ref_chunks - cmp_chunks))
+                if args.json:
+                    dump_json({
+                        "path": path,
+                        "what": "content",
+                        "added": added,
+                        "removed": removed,
+                        "total_reference": total_ref,
+                        "total_compare": total_compare,
+                    })
+                else:
+                    print(remove_surrogates(path), "different contents")
+                    print("\t+%s, -%s, ref %s, compare %s" % (format_bytes(added), format_bytes(removed),
+                                                              format_bytes(total_ref), format_bytes(total_compare)))
+
+        repository = self.open_repository(args)
+        manifest, key = Manifest.load(repository)
+        ref = Archive(repository, key, manifest, args.location.archive)
+        compare = Archive(repository, key, manifest, args.compare)
+
+        # Extract and check chunker params
+        cmdline_parser = self.build_parser()
+        ref_chunker_params = ref.metadata.get(b'chunker_params', cmdline_parser.parse_args(
+            ref.metadata[b'cmdline'][1:]).chunker_params)
+        cmp_chunker_params = compare.metadata.get(b'chunker_params', cmdline_parser.parse_args(
+            compare.metadata[b'cmdline'][1:]).chunker_params)
```

Comment:
keep it simple. use reliable information and if there is none, do it in a safe way.

the issue with old archives being slow then will solve automatically over time. ;)